### ISSUE TITLE
Intermittent lag in "Malmo" map, supposedly due to the AI

### ISSUE THREAD

**Issue by akien-mga** [Andrettin/Wyrmgus#30]
As seen in the attached saved game [1], there is intermittent lag (every 15-20s or something, though I'm not sure it's perfectly regular) in the Malmo map as player in the 5th scenario of the germanic campaign.

Some pointers to identify the source of the lag, as discussed on chat:
```
Akien: It's the first map where I experience lags
Akien: It did not lag at first though, only after I had explored half of the map I'd say
Andrettin: hmm, it's not an issue related to the scenario triggers, I just tested that
Andrettin: it doesn't seem like it has anything to do with the time of the day changes either, as the lag is not happening when the time of the day changes
Andrettin: maybe the lag comes from the AI reassigning its workers to gather a different resources
Akien: Would that happen often?
Andrettin: no, it shouldn't happen often
Andrettin: and I'm not seeing any indication of it happening often either
Andrettin: hmm a debug executable doesn't tell me anything about it either
Andrettin: killing all AI workers and their town hall does seem to get rid of the lag
Andrettin: I imagine the cause is this function:
Andrettin: AiAssignHarvesterFromUnit (ai_resource.cpp)
Andrettin: it is pretty expensive performance-wise, and is probably being called more often than it should
Andrettin: and the recent additions I made to check if there is an equivalent resource (like silver for gold) probably impacted performance further
```

[1] http://dl.free.fr/kKlGFYth5

**Comment by Andrettin** [comment]
Fixed in https://github.com/Andrettin/Wyrmgus/commit/321bf2f2a029845d61f31f249710115d4f467b88
and https://github.com/Andrettin/Wyrmgus/commit/130c7dc99fa543d54cceee3607261be6ea06ca6b

It seems that the problem was the following: the AI wanted to gather stone, so it was trying to reassign its workers to do so. However, since there was no stone in the map, no worker actually got reassigned to do that, so the algorithm kept trying to reassign the workers to gather stone. This in itself wouldn't have been a problem - the issue is that the algorithm recalculated the search for stone for every worker it tried to reassign, which was every worker that was gathering any other resource. I fixed the algorithm so that now it doesn't recalculate the resource search for workers of the same type.
